MappingPGeneric

peterrum opened the pull request:
References https://github.com/dealii/dealii/pull/10646.

The class `FEValues::reinit()` needs for a face (see https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/source/fe/fe_values.cc#L4809-L4835):

- [`FiniteElement::get_data()`](https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/include/deal.II/fe/fe_poly_tensor.h#L236-L376) -> `FiniteElement::InternalDataBase`
- `Mapping::get_data()` -> `Mapping::InternalDataBase`
- [`FiniteElement::fill_fe_values()`](https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/source/fe/fe_poly_tensor.cc#L368-L939), [`FiniteElement::fill_fe_face_values()`](https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/source/fe/fe_poly_tensor.cc#L943-L1564), [`FiniteElement::fill_fe_subface_values()`](https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/source/fe/fe_poly_tensor.cc#L1568-L2182) -> `FiniteElementRelatedData`
- `Mapping::fill_fe_values()`, `Mapping::fill_fe_face_values()`, `Mapping::fill_fe_subface_values()` -> `MappingRelatedData`

drwells committed:
Fix doxygen underscores.

We need to use $\mathtt{A\_B}$ and not $\texttt{A\_B}$ to correctly print an
underscore and not the backslash.

nfehn committed:
fix bug in fast version of GridTools::find_active_cell_around_point()

Rombur committed:
Fix typo

Rombur committed:
Replace macro with constexpr variable

tjhei committed:
cleanup a confusing cmake doxygen message

rezarastak committed:
DiscreteTime::set_next_step_size() added

masterleinad committed:
Remove deprecated DataOutBase::tecplot_binary

luca-heltai committed:
Added unit_to_real and real_to_unit to BoundingBox

peterrum committed:
Refactor Triangulation::create_triangulation()

masterleinad committed:
Remove constraint_and_return_value and types_are_equal

peterrum committed:
Generalize line_orientation() for 2D general meshes

peterrum committed:
Introduce ReferenceCell::Type

peterrum committed:
Introduce ReferenceCell::Type

peterrum committed:
Remove GeometryInfro from a few files

tamiko committed:
examples/step-69: Link in repository for higher-order code

peterrum committed:
Remove some GeometryInfos from dof_accessor.h

peterrum committed:
Intoduce ReferenceCellInfo

nfehn committed:
fix tests

peterrum committed:
Add Simplex::ScalarPolynomial

peterrum committed:
Add additional quadrature rule for Simplex

masterleinad committed:
Fix default mapping parameter in VectorTools boundary functions

peterrum committed:
Replace dofs_per_vertex -> n_dofs_per_vertex()

Replace dofs_per_cell -> n_dofs_per_cell()

peterrum committed:
Intoduce MappingFE

marcfehling committed:
Bugfix: Unify dof indices on ghost interfaces if no finite element dominates.

By design, there will be always one dominating element for vertices.

peterrum commented on include/deal.II/fe/mapping_p_generic.h:
```diff
@@ -134,11 +134,9 @@ class MappingPGeneric : public Mapping<dim, spacedim>
+   * TODO
    */
-  MappingPGeneric(const unsigned int polynomial_degree);
+  MappingPGeneric(const FiniteElement<dim, spacedim> &fe);
```
Build around a `FiniteElement` object.

peterrum commented on source/fe/mapping_p_generic.cc:
```diff
@@ -3824,275 +2111,20 @@ MappingPGeneric<dim, spacedim>::transform(
+  std::vector<Point<spacedim>> a(cell->n_vertices());
+
+  for (const unsigned int i : cell->vertex_indices())
+    a[i] = cell->vertex(i);
```
How should we extend the method `MappingPGeneric<dim, spacedim>::compute_mapping_support_points()` to higher degrees? Might be related to https://github.com/dealii/dealii/pull/10163.

drwells replied:
I don't see how it is related to #10163. As far as that patch goes - I would prefer to load the data into a finite element field and use that instead of relying on an ad-hoc mapping class. If we got that to work from VTK input then we would automatically support this case too.

drwells replied:
We should definitely not worry about higher-order yet since there is no way to generalize Gauss-Lobatto to simplices. I suspect the higher-order solution will be hierarchical, not nodal - if I recall correctly nek uses hierarchical finite elements.

peterrum commented on source/fe/mapping_p_generic.cc:
```diff
@@ -664,975 +91,181 @@ MappingPGeneric<dim, spacedim>::InternalData::initialize(
-              invert(df) * static_cast<const Tensor<1, dim> &>(f);
-            p_unit -= d;
+      // Compute tangentials to the unit cell.
+      for (const unsigned int i : GeometryInfo<dim>::face_indices())
+        {
+          unit_tangentials[i].resize(n_original_q_points);
+          std::fill(unit_tangentials[i].begin(),
+                    unit_tangentials[i].end(),
+                    GeometryInfo<dim>::unit_tangential_vectors[i][0]);
+          if (dim > 2)
+            {
+              unit_tangentials[GeometryInfo<dim>::faces_per_cell + i].resize(
+                n_original_q_points);
+              std::fill(
+                unit_tangentials[GeometryInfo<dim>::faces_per_cell + i].begin(),
+                unit_tangentials[GeometryInfo<dim>::faces_per_cell + i].end(),
+                GeometryInfo<dim>::unit_tangential_vectors[i][1]);
+            }
+        }
+    }
```
Needs to depend on the reference cell.

peterrum reviewed (commented):
Needed changes.

Further question: how to deal with
- [`QProjector<dim>::DataSetDescriptor::face`](https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/source/base/quadrature.cc#L1137-L1203)
- [`QProjector<dim>::DataSetDescriptor::subface`](https://github.com/dealii/dealii/blob/08727cc441b9cfd1081cb677fb75e3b62c8ba060/source/base/quadrature.cc#L1207-L1498)

drwells commented on include/deal.II/fe/mapping_p_generic.h:
```diff
@@ -252,7 +254,7 @@ class MappingPGeneric : public Mapping<dim, spacedim>
      * to which this object will correspond.
      */
-    InternalData(const unsigned int polynomial_degree);
+    InternalData(const FiniteElement<dim, spacedim> &fe);
```
Is this a `FiniteElement` so that you know the element geometry?

peterrum replied:
Yes: polynomials + geometric shape.

drwells commented on include/deal.II/fe/mapping_p_generic.h:
```diff
@@ -421,6 +423,11 @@ class MappingPGeneric : public Mapping<dim, spacedim>
+    /**
+     * TODO
+     */
+    const FiniteElement<dim, spacedim> fe;
```
should probably be a smart pointer (this isn't a big deal at this stage)

peterrum replied:
I was thinking if it wasn't better to clone it and put it into a unique pointer!?

drwells commented:
Here is the 'original' version of DataSetDescriptor::subface:
```cpp
template <int dim>
typename QProjector<dim>::DataSetDescriptor
QProjector<dim>::DataSetDescriptor::
subface (const unsigned int face_no,
         const unsigned int subface_no,
         const bool         face_orientation,
         const bool         face_flip,
         const bool         face_rotation,
         const unsigned int n_quadrature_points)
{
  Assert (dim != 1, ExcInternalError());
  Assert (face_no < GeometryInfo<dim>::faces_per_cell,
          ExcInternalError());
                                   // the trick with +1 prevents that we get a
                                   // warning in 1d
  Assert (subface_no+1 < GeometryInfo<dim>::subfaces_per_face+1,
          ExcInternalError());
  
				   // in 3d, we have to account for faces that
				   // have non-standard face orientation, flip
				   // and rotation. thus, we have to store
				   // _eight_ data sets per face or subface
  
				   // set up a table with the according offsets
				   // for non-standard orientation, first index:
				   // face_orientation (standard true=1), second
				   // index: face_flip (standard false=0), third
				   // index: face_rotation (standard false=0)
				   //
				   // note, that normally we should use the
				   // obvious offsets 0,1,2,3,4,5,6,7. However,
				   // prior to the changes enabling flipped and
				   // rotated faces, in many places of the
				   // library the convention was used, that the
				   // first dataset with offset 0 corresponds to
				   // a face in standard orientation. therefore
				   // we use the offsets 4,5,6,7,0,1,2,3 here to
				   // stick to that (implicit) convention
  static const unsigned int offset[2][2][2]=
    {{
					   // face_orientation=false; face_flip=false; face_rotation=false and true
	  {4*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face,
	   5*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face},
					   // face_orientation=false; face_flip=true;  face_rotation=false and true
	  {6*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face,
	   7*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face}},   
     {
					    // face_orientation=true;  face_flip=false; face_rotation=false and true
	   {0*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face,
	    1*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face},
					    // face_orientation=true;  face_flip=true;  face_rotation=false and true
	   {2*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face,
	    3*GeometryInfo<dim>::faces_per_cell*GeometryInfo<dim>::subfaces_per_face}}};
  
  switch (dim)
    {
      case 1:
      case 2:
            return ((face_no * GeometryInfo<dim>::subfaces_per_face +
                     subface_no)
                    * n_quadrature_points);
      case 3:
            return (((face_no * GeometryInfo<dim>::subfaces_per_face +
                      subface_no)
                     + offset[face_orientation][face_flip][face_rotation]
                     )
                    * n_quadrature_points);
      default:
            Assert (false, ExcInternalError());
    }
  return numbers::invalid_unsigned_int;              
}
```

drwells commented:
See commit b3a7efd

drwells commented:
```cpp
template <int dim>
typename QProjector<dim>::DataSetDescriptor
QProjector<dim>::DataSetDescriptor::
face (const unsigned int face_no,
      const bool         face_orientation,
      const bool         face_flip,
      const bool         face_rotation,
      const unsigned int n_quadrature_points)
{
  Assert (dim != 1, ExcInternalError());
  Assert (face_no < GeometryInfo<dim>::faces_per_cell,
          ExcInternalError());
  
				   // in 3d, we have to account for faces that
				   // have non-standard face orientation, flip
				   // and rotation. thus, we have to store
				   // _eight_ data sets per face or subface

				   // set up a table with the according offsets
				   // for non-standard orientation, first index:
				   // face_orientation (standard true=1), second
				   // index: face_flip (standard false=0), third
				   // index: face_rotation (standard false=0)
				   //
				   // note, that normally we should use the
				   // obvious offsets 0,1,2,3,4,5,6,7. However,
				   // prior to the changes enabling flipped and
				   // rotated faces, in many places of the
				   // library the convention was used, that the
				   // first dataset with offset 0 corresponds to
				   // a face in standard orientation. therefore
				   // we use the offsets 4,5,6,7,0,1,2,3 here to
				   // stick to that (implicit) convention
  static const unsigned int offset[2][2][2]=
    {{{4*GeometryInfo<dim>::faces_per_cell, 5*GeometryInfo<dim>::faces_per_cell},    // face_orientation=false; face_flip=false; face_rotation=false and true
      {6*GeometryInfo<dim>::faces_per_cell, 7*GeometryInfo<dim>::faces_per_cell}},   // face_orientation=false; face_flip=true;  face_rotation=false and true
     {{0*GeometryInfo<dim>::faces_per_cell, 1*GeometryInfo<dim>::faces_per_cell},    // face_orientation=true;  face_flip=false; face_rotation=false and true
      {2*GeometryInfo<dim>::faces_per_cell, 3*GeometryInfo<dim>::faces_per_cell}}};  // face_orientation=true;  face_flip=true;  face_rotation=false and true

  switch (dim)
    {
      case 1:
      case 2:
            return face_no * n_quadrature_points;

	    
      case 3:
            return ((face_no
		     + offset[face_orientation][face_flip][face_rotation])
		    * n_quadrature_points);

      default:
            Assert (false, ExcInternalError());
    }
  return numbers::invalid_unsigned_int;
}
```

tjhei committed:
Merge pull request #10643 from peterrum/simplex_quadrature

Add additional quadrature rule for Simplex

peterrum committed:
Merge pull request #10640 from tamiko/step-69-references

examples/step-69: Link in repository for higher-order code

masterleinad committed:
Remove deprecated TransposeTableIterators

masterleinad committed:
Remove deprecated CutOffFunctioNbase member functions

masterleinad committed:
Remove Solver

masterleinad committed:
Remove PointValueHostory::get_point

dangars committed:
Replace MPI_COMM_WORLD by mpi_communicator

tjhei committed:
fix warning in bundled taskflow

fixes ``warning: extra ‘;’ [-Wpedantic]``

masterleinad committed:
Merge pull request #10657 from dangars/step_50_mpi_communicator

Replace MPI_COMM_WORLD by mpi_communicator in step-50

masterleinad committed:
Merge pull request #10658 from tjhei/taskflow-warning-fix

fix warning in bundled taskflow

peterrum committed:
Merge pull request #10655 from masterleinad/more_deprecatiosn

Remove more deprecations

peterrum committed:
Merge pull request #10654 from masterleinad/remove_transpose_table_iterators

Remove deprecated TransposeTableIterators

peterrum committed:
Merge pull request #10647 from masterleinad/fix_default_vector_tools_boundary

Fix default mapping parameter in VectorTools boundary functions

peterrum committed:
Merge pull request #10636 from masterleinad/remove_template_constraints

Remove constraint_and_return_value and types_are_equal

masterleinad committed:
Remove ResidualSimple::initialize_local_blocks

masterleinad committed:
Remove MGConstrainedDoFs::get_level_constraint_matrix

tjhei committed:
hide instantiations in Utilities from doxygen

If you look at
https://www.dealii.org/developer/doxygen/deal.II/namespaceUtilities_1_1MPI.html
you see several pages of instantiations of sum/min/max/etc. without
documentation.

Similary, Utilities contains pack_integers,
inverse_Hilbert_space_filling_curve, and other instantiations.

Hide these from doxygen.

Rombur committed:
Fix deprecation warning

Rombur committed:
Fix test

Rombur committed:
Add get_vector_partitioner and get_dof_handler functions

Rombur committed:
Add changelog entry

masterleinad committed:
Remove numbers::invalid_manifold_id

bangerth committed:
Add warnings to functions that set particle properties.

jppelteret committed:
Use neutral terminology where possible

peterrum committed:
Merge pull request #10663 from bangerth/doc-99

Add warnings to functions that set particle properties.

peterrum committed:
Add some consts in step-12

tjhei committed:
teach expand_instantiations to deal with preprocessor defines

tjhei committed:
Merge pull request #10664 from peterrum/step12_consts

Add some consts in step-12

masterleinad committed:
Merge pull request #10651 from marcfehling/unification-without-domination

Bugfix: Unify dof indices on ghost interfaces if no finite element dominates.

nfehn committed:
fix another test (mpi/fe_field_function_03)

nfehn committed:
fix another test (particles/generators_07)

drwells committed:
Merge pull request #10635 from peterrum/line_orientation_general

Generalize line_orientation() for 2D general meshes

drwells committed:
Merge pull request #10622 from masterleinad/remove_tecplot_binary

Remove deprecated DataOutBase::tecplot_binary

drwells committed:
Merge pull request #10253 from drwells/fix-doxygen-underscores

Fix doxygen underscores.

drwells committed:
Merge pull request #10630 from luca-heltai/bbox-transformations

Added unit_to_real and real_to_unit to BoundingBox

drwells committed:
Merge pull request #10662 from masterleinad/deprecations_2

Remove more deprecations

drwells committed:
Merge pull request #10613 from tjhei/cleanup_cmake_doxygen_message

cleanup a confusing cmake doxygen message

tjhei committed:
reshuffle namespaces in workstream class

No big change in functionality, but needed for the transition to
taskflow:
- extract the TBB implementation into a separate namespace
- extract the sequential version
- sequential version caches empty worker/copier check
- add a few comments

part of #10391

tjhei committed:
run indent on expand_instantiations.cc

Also run clang-format on cmake/scripts/*.cc.

While I am here, add "tests" to remove_trailing_whitespace that was
missing.

drwells committed:
Add some template instantiations that were accidentally removed.

masterleinad committed:
Fix trilinos/68

drwells committed:
Merge pull request #10660 from tjhei/doxygen_utilities_hide_instantiations

hide instantiations in Utilities from doxygen

masterleinad committed:
Merge pull request #10642 from peterrum/simplex_polynomials

Add Simplex::ScalarPolynomial

peterrum committed:
Generalize face_orientation() for 2D general meshes

peterrum committed:
Be explicit regarding mapping in step-12

drwells committed:
Merge pull request #10639 from peterrum/gi_1

Remove GeometryInfro from a few files

drwells committed:
Merge pull request #10637 from peterrum/geometric_entity_type

Introduce ReferenceCell::Type

drwells committed:
Merge pull request #10668 from masterleinad/fix_trilinos_68

Fix trilinos/68

drwells committed:
Merge pull request #10670 from peterrum/step12-mapping

Be explicit regarding mapping in step-12

drwells committed:
Merge pull request #10669 from peterrum/face_orientation_general

Generalize face_orientation() for 2D general meshes

drwells committed:
Merge pull request #10667 from drwells/readd-boundary-projection-instantiations

Add some template instantiations that were accidentally removed.

masterleinad committed:
Merge pull request #10666 from tjhei/format-expand-inst

run indent on expand_instantiations.cc

kronbichler committed:
Merge pull request #10540 from nfehn/bug_fix_find_active_cells

fix bug in fast version of GridTools::find_active_cell_around_point()

peterrum committed:
Merge pull request #10661 from Rombur/cuda_mf_dof_handler

Add get_vector_partitioner and get_dof_handler to CUDAWrappers::MatrixFree

masterleinad committed:
Remove deprecated Gnuplot::n_boundary_face_points

tjhei committed:
fix taskflow quicktest

The macro was renamed and we forgot to change this.

tjhei committed:
fix colored sequential code

luca-heltai committed:
Fixed mesh_loop in one dimension.

tjhei committed:
add test

kronbichler committed:
Merge pull request #10641 from peterrum/GeometricEntityInfo

Intoduce ReferenceCellInfo

masterleinad committed:
Remove deprecated MGTools::extract_non_interface_dofs

peterrum committed:
Use ReferenceCell::Type to get Info

peterrum committed:
Use ReferenceCell::Type to get Info

Add additional ReferenceCell::internal::Infos

tjhei committed:
extract TBB colored WorkStream::run

tjhei committed:
update test

tjhei committed:
add missing inline

masterleinad committed:
Instantiate DoFTools::make_boundary_sparsity_pattern for complex types

masterleinad committed:
Remove lac/parallel_vector.h

masterleinad committed:
Remove redundant constructors

luca-heltai committed:
Be explicit about dimension.

tjhei committed:
Merge pull request #10672 from luca-heltai/fix-meshloop-1d

Fixed mesh_loop in one dimension.

masterleinad committed:
Fix compiler warnings for old clang

peterrum committed:
Add check to step-12

peterrum committed:
Implement FEInterfaceValues::get_mapping() and ::get_fe

tjhei committed:
Merge pull request #10681 from peterrum/feinterfacecalues_mapping_fe

Implement FEInterfaceValues::get_mapping() and ::get_fe

peterrum committed:
Fix typo

peterrum committed:
Merge pull request #10678 from masterleinad/fix_warning_old_clang

Fix compiler warnings for old clang

bangerth committed:
Update the readme for C++14.

tjhei committed:
Merge pull request #10683 from bangerth/readme

Update the readme for C++14.

tjhei committed:
doc: fix incorrect default for DEAL_II_ALLOW_PLATFORM_INTROSPECTION

bangerth committed:
Merge pull request #10684 from tjhei/doc_platform_introspection

doc: fix incorrect default for DEAL_II_ALLOW_PLATFORM_INTROSPECTION

peterrum committed:
Remove many GeometryInfos from grid/grid_tools

peterrum committed:
Merge pull request #10675 from masterleinad/implement_complex_dof_tools

Instantiate DoFTools::make_boundary_sparsity_pattern for complex types

peterrum committed:
Merge pull request #10676 from masterleinad/deprecations_3

Remove deprecated functions

peterrum committed:
Remove some GeometryInfos from dofs/dof_tools

peterrum committed:
Merge pull request #10682 from peterrum/reference_cell_typo

Fix typo

peterrum committed:
Merge pull request #10680 from peterrum/step12-check

Add check to step-12

peterrum committed:
Add Simplex::GridGenerator

peterrum committed:
Remove some GeometryInfos from grid/tria_accessor.cc

peterrum committed:
Implement Simplex::ScalarPolynomial and Simplex::PGauss also for 1D

Reorder shape functions

peterrum committed:
Merge pull request #10671 from tjhei/fix_taskflow_quicktest

fix taskflow quicktest and fix taskflow linker error

luca-heltai committed:
Merge pull request #10688 from peterrum/simplex_gridgenerator

Add Simplex::GridGenerator

peterrum committed:
Generalize GridOut::write_vtk()

masterleinad committed:
Merge pull request #10549 from jppelteret/update_nomenclature_01

Use neutral terminology where possible

masterleinad committed:
Remove deprecated MGConstrainedDoFs::initialize overload

peterrum committed:
Remove some if-statements

masterleinad committed:
Merge pull request #10696 from peterrum/gi_remove_checks

Remove some if-statements

peterrum committed:
Fix a bug in Simplex::GridGenerator

masterleinad committed:
Merge pull request #10697 from peterrum/simplex_gridgenerator_bug_fix

Fix a bug in Simplex::GridGenerator

masterleinad committed:
Merge pull request #10693 from peterrum/simplex_gridou_write_vtk

Generalize GridOut::write_vtk()

peterrum committed:
Remove many GeometricInfo usages from examples

peterrum committed:
Add simplex module page

masterleinad committed:
Fix GeneralDataStorage::get_[or_add_]object_with_name for ICC

tjhei committed:
Merge pull request #10700 from masterleinad/fix_intel_general_data_storage

Fix GeneralDataStorage::get_[or_add_]object_with_name for ICC

masterleinad committed:
Remove manual additions to LAPACK templates

masterleinad committed:
Merge pull request #10702 from masterleinad/remove_deprecated_lapack_templates

Remove manual additions to LAPACK templates

drwells committed:
Merge pull request #10689 from peterrum/gi_triaaccessor

Remove some GeometryInfos from grid/tria_accessor.cc

peterrum committed:
Fix cast in communicate_active_fe_indices()

drwells committed:
Merge pull request #10685 from peterrum/gi_gridtools

Remove many GeometryInfos from grid/grid_tools

drwells committed:
Merge pull request #10704 from peterrum/communicate_active_fe_indices

Fix cast in communicate_active_fe_indices()

peterrum committed:
Add new hp::FEFaceValues::reinit() functions

peterrum committed:
Add new hp::MappingCollection constructor

peterrum committed:
hp::FEValues::reinit() pass be reference

peterrum committed:
Accept MappingCollection in DataOut::build_patches()

masterleinad committed:
Merge pull request #10706 from peterrum/hp_fe_values_reinit_reference

hp::FEValues::reinit() pass by reference

peterrum committed:
Replace some GeometryInfos in dof_handler.cc and dof_handler_policy.cc

peterrum committed:
Fix dynamic cast

drwells committed:
Merge pull request #10686 from peterrum/gi_doftools

Remove some GeometryInfos from dofs/dof_tools

masterleinad committed:
Merge pull request #10707 from peterrum/hp_mappingcollection_constructor

Add new hp::MappingCollection constructor

masterleinad committed:
Merge pull request #10710 from peterrum/dofhandler_cast

Fix dynamic cast

bangerth committed:
Merge pull request #10709 from peterrum/gi_dofhandlerpolicy

Replace all GeometryInfos in dof_handler.cc and dof_handler_policy.cc

bangerth committed:
Merge pull request #10665 from tjhei/tbb-refactor-namespace

reshuffle namespaces in workstream class

bangerth committed:
Merge pull request #10679 from tjhei/tbb-refactor-colored

extract TBB colored WorkStream::run

bangerth committed:
Avoid the use of raw pointers in WorkStream.

We've tried to not use raw pointers throughout the library, but WorkStream had
remaining places. This is not necessary -- it just requires moving a couple of
lines of code around.

peterrum committed:
Use more n_dofs_per_ and add functions for offsets

tjhei committed:
Merge pull request #10711 from bangerth/workstream

Avoid the use of raw pointers in WorkStream.

drwells committed:
Merge pull request #10699 from peterrum/simplex_module_page

Add simplex module page

tjhei committed:
detect inifinite loop in periodic refinement

As witnessed by ASPECT, see
https://github.com/geodynamics/aspect/issues/3604 ,
we might end up in an infinite loop while trying to prepare refinement
flags in
parallel::distributed::Triangulation::prepare_coarsening_and_refinement().
Add a check and abort (instead of a hang).

drwells committed:
Merge pull request #10645 from peterrum/fe_base_gi

FiniteElementData: select number of objects based on cell type

drwells committed:
Merge pull request #10674 from peterrum/dofs_per_function_1

Replace FiniteElementData::dofs_per_* with FiniteElementData::n_dofs_per_*

luca-heltai committed:
Merge pull request #10698 from peterrum/examples_gi

Remove many GeometryInfo usages from examples

kronbichler committed:
Fix typo in computation of speed of sound

kronbichler committed:
Merge pull request #10712 from tjhei/periodic_hang_abort

detect infinite loop in periodic refinement

kronbichler committed:
Merge pull request #10694 from masterleinad/remove_deprecated_mgconstraineddofs_initialize

Remove deprecated MGConstrainedDoFs::initialize overload

kronbichler committed:
Merge pull request #10631 from peterrum/creat_triangulation

Refactor Triangulation::create_triangulation()

kronbichler committed:
Merge pull request #10673 from peterrum/use_reference_cell_type

Use ReferenceCell::Type to get Info

drwells committed:
Merge pull request #10692 from peterrum/additional_cell_types

Add additional ReferenceCell::internal::Infos

bangerth committed:
Merge pull request #10713 from kronbichler/doc-step-67

[step-67] Fix typo in computation of speed of sound

peterrum committed:
Add missing CRS entry and some asserts

peterrum committed:
Unitfy set/get_dof_indices() and set/get_mg_dof_indices()

peterrum committed:
Store reference cell type in FiniteElementData

peterrum committed:
Remove many GeometryInfos from fe_tools.templates.h

peterrum committed:
Move declaration of explicit specializations to correct file

peterrum committed:
Intoduce ReferenceCell in QProjector

Rombur committed:
Merge pull request #10719 from peterrum/qiterated_file

Rombur committed:
Merge pull request #10690 from peterrum/simplex_poly_quad_1D

Implement Simplex::ScalarPolynomial and Simplex::PGauss also for 1D ...

bangerth committed:
Fix a typo.

masterleinad committed:
Merge pull request #10721 from bangerth/typo

Fix a typo.

kronbichler committed:
Merge pull request #10705 from peterrum/fe_face_values_reinit

Add new hp::FEFaceValues::reinit() functions

kronbichler committed:
Merge pull request #10714 from peterrum/dofhandler_missing_crs_entry

Add missing CRS entry and some asserts

tjhei committed:
fix hang in periodic mesh smoothing flag

The mesh smoothing flag eliminate_refined_boundary_islands creates an
infinite loop if it detects such a cell at a periodic boundary that
requires refinement due to the periodic 2:1 rule.
Work around this by not treating cells at periodic boundaries as
"refined boundary islands".

tjhei committed:
add test

peterrum committed:
Implement TriaAccessor::measure() and TriaAccessor::diameter() for simplex

peterrum committed:
Generalize FETools::Compositing::multiply_dof_numbers

peterrum committed:
Fix typo

luca-heltai committed:
Merge pull request #10728 from peterrum/eliminate_refined_inner_islands_typo

Fix typo

drwells committed:
Merge pull request #10621 from rezarastak/set_time_step_size

DiscreteTime::set_next_step_size() added

kronbichler committed:
Merge pull request #10708 from peterrum/data_out_mapping_collection

Accept MappingCollection in DataOut::build_patches()

kronbichler committed:
Merge pull request #10718 from peterrum/gi_fe_tools

Remove many GeometryInfos from fe_tools.templates.h

kronbichler committed:
Merge pull request #10716 from peterrum/set_get_mg_dof_indices

Unitfy set/get_dof_indices() and set/get_mg_dof_indices()

tjhei committed:
update comment

drwells committed:
Fix a warning about virtual destructors and virtual functions.

drwells committed:
Merge pull request #10726 from peterrum/fetools_compositing_multiply_dof_numbers

Generalize FETools::Compositing::multiply_dof_numbers

drwells committed:
Merge pull request #10725 from peterrum/simplex_cellmeasure_diameter

Implement TriaAccessor::measure() and ::diameter() for simplex

peterrum committed:
Simplify GridTools::get_face_connectivity_of_cells

peterrum committed:
Merge pull request #10729 from drwells/dtor-warning

Fix a warning about virtual destructors and virtual functions.

kronbichler committed:
Merge pull request #10723 from tjhei/periodic_smoothen_hang_fix

fix hang in periodic mesh smoothing flag

kronbichler committed:
Merge pull request #10732 from peterrum/get_face_connectivity_of_cells

Simplify GridTools::get_face_connectivity_of_cells

kronbichler committed:
Merge pull request #10720 from peterrum/qprojection

Pass ReferenceCell::Type to QProjection
